From 2507c32c7910aa6049a8de443e2f7e2a28dec5ec Mon Sep 17 00:00:00 2001
From: Mark Fasheh <mfasheh@suse.com>
Date: Thu, 1 Sep 2011 15:34:34 -0700
Subject: [PATCH 15/20] btrfs: Go readonly on missing ref in
 btrfs_get_parent()
Patch-mainline: 3.2?

btrfs_get_parent() searches the btree for a ref to the current object. From
there it can compute the parent objectid from which it can return a dentry.
if the reference is not found in the tree however, we BUG(). I believe a
more appropriate response would be to go read-only as this seems like a
corruption.

Signed-off-by: Mark Fasheh <mfasheh@suse.com>
---
 fs/btrfs/export.c |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/fs/btrfs/export.c b/fs/btrfs/export.c
index 1b8dc33..a335169 100644
--- a/fs/btrfs/export.c
+++ b/fs/btrfs/export.c
@@ -193,7 +193,13 @@ static struct dentry *btrfs_get_parent(struct dentry *child)
 	if (ret < 0)
 		goto fail;
 
-	BUG_ON(ret == 0);
+	if (ret == 0) {
+		/* Missing ref, should go readonly. */
+		ret = -ENOENT;
+		btrfs_std_error(root->fs_info, ret);
+		goto fail;
+	}
+
 	if (path->slots[0] == 0) {
 		ret = -ENOENT;
 		goto fail;
-- 
1.7.6

